core/object/worker_thread_pool.cpp

```diff
@@ -72,29 +96,47 @@ void WorkerThreadPool::_process_task(Task *p_task) {
 	if (p_task->group) {
 		// Handling a group
 		bool do_post = false;
-		Callable::CallError ce;
-		Variant ret;
-		Variant arg;
-		Variant *argptr = &arg;
 
-		while (true) {
-			uint32_t work_index = p_task->group->index.postincrement();
+		// Process nearby indices in the same thread for better caching
+		const uint32_t batch_size = calculate_batch_size(p_task->group->max, p_task->group->tasks_used);
 
+		while (true) {
+			uint32_t work_index = p_task->group->index.postadd(batch_size);
 			if (work_index >= p_task->group->max) {
 				break;
 			}
+
+#ifdef DEBUG_ENABLED
```

Comment:
```suggestion
#ifdef DEV_ENABLED
```

And I think I'd have an SCons build flag (or just a commented out define at the top of the file) so this can be opted-out.

Reply:
> And I think I'd have an SCons build flag (or just a commented out define at the top of the file) so this can be opted-out.

If the intent is for debugging purposes, would setting `threading/worker_pool/max_threads = 1` (or 0) suffice? This avoids the need for a rebuild, and can even be done from the command line via environment variable support: #72689

---

In general, I would like to avoid creating a false sense of security where code appears to work during engine/game development, only for players to experience inexplicable non-reproducible bugs that were actually caused by race conditions.

Ideally, threading should have a simple interface where setting `threading/worker_pool/max_threads = 0 or 1` results in stable race-free behavior, and other values result in faster code but with any race conditions consistently occurring (rather than lurking as heisenbugs). `DEBUG_ENABLED` means that users will also experience them, and therefore be able to help by filing bug reports. (#72577 was opened under a similar reasoning)

As a side note, since `WorkerThreadPool` can be reached from GDScript, this can also help users debug their multithreading code, without them needing to compile a full `use_tsan=yes` build.